File: image_signal/report_7.py
import cv2
import numpy as np
import time


def expand_size(tr_matrix, size=(0, 0)):
    x, y = size
    vertices = np.array([[[0], [0]], [[x], [0]], [[0], [y]], [[x], [y]]])
    x_ = []
    y_ = []
    for vertex in vertices:
        vertex_trans = np.dot(tr_matrix, vertex)
        x_.append(vertex_trans[0][0])
        y_.append(vertex_trans[1][0])
    x_min = round(min(x_))
    y_min = round(min(y_))
    x_ = round(max(x_)) - x_min
    y_ = round(max(y_)) - y_min
    return np.int(x_), np.int(y_), x_min, y_min


img = cv2.imread('tokyoskytree.jpg', 0)
theta = -np.pi/6
tr = np.matrix([[np.cos(theta), -np.sin(theta)], [np.sin(theta), np.cos(theta)]])


def trans_in(img, tr):
    tr_in = tr**-1
    x, y, shift_x, shift_y = expand_size(tr, img.shape)
    img_tr = np.zeros((x, y), dtype='uint8')
    for i in range(img_tr.shape[0]):
        for j in range(img_tr.shape[1]):
            co = np.array([[i + shift_x], [j + shift_y]])
            co = np.array(np.dot(tr_in, co))
            i_ = int(round(co[0][0]))
            j_ = int(round(co[1][0]))
            try:
                if i_ < 0 or j_ < 0:
                    raise IndexError
                img_tr[i][j] = img[i_][j_]
            except IndexError:
                pass
    return img_tr


def trans(img, tr):
    x, y, shift_x, shift_y = expand_size(tr, img.shape)
    img_tr = np.zeros([x, y], dtype='uint8')
    for i in range(img.shape[0]):
        for j in range(img.shape[1]):
            co = np.array(np.dot(tr, np.array([[i], [j]])))
            i_, j_ = co[0][0] - shift_x, co[1][0] - shift_y
            try:
                img_tr[i_][j_] = img[i][j]
            except IndexError:
                pass
    return img_tr


def trans_inverse(img, tr_inverse, shift=(0, 0)):
    img2 = np.zeros((img.shape[0]*2, img.shape[1]*2), dtype='uint8')
    for i in range(img2.shape[0]):
        for j in range(img2.shape[1]):
            co = np.dot(tr_inverse, np.array([[i + shift[0]], [j + shift[1]]]))
            try:
                img2[i][j] = img[np.int(co[0][0])][np.int(co[1][0])]
            except IndexError:
                img2[i][j] = 255
    return img2


start = time.clock()
img2 = trans_in(img, tr)
img3 = trans_inverse(img2, np.matrix([[2,0],[0,2]])**-1)
print('time = %f s' % (time.clock() - start))

cv2.imshow('or', img)
cv2.imshow('tr_in', img2)
cv2.imshow('tr', img3)
cv2.waitKey(0)

File: image_signal/report_7.cpp
// opencv 3.0
// g++ `pkg-config --cflags --libs opencv`

#include <iostream>
#include <math.h>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

// Matrix multiplicative of R2x2 * R2x1
int *mat2x2(float R[2][2], int co[2]){
  static int co_[2];
  co_[0] = (int)(R[0][0]*co[0] + R[0][1]*co[1] + 0.5);
  co_[1] = (int)(R[1][0]*co[0] + R[1][1]*co[1] + 0.5);
  return co_;
}

// calculate size of translated image
int *expand_size(float R[2][2], int x, int y){
  int *size;
  int i = 0;
  int vertices[4][2] = {{0, 0}, {x, 0}, {0, y}, {x, y}};
  size = mat2x2(R, vertices[0]);
  double max_x, min_x, max_y, min_y;
  max_x = min_x = size[0];
  max_y = min_y = size[1];
  for(i=1; i<4; i++){
    size = mat2x2(R, vertices[i]);
    if (size[0] > max_x)
      max_x = size[0];
    if (size[0] < min_x)
      min_x = size[0];
    if (size[1] > max_y)
      max_y = size[1];
    if (size[1] < min_y)
      min_y = size[1];
  }

  int ma_x,ma_y,mi_x,mi_y;
  mi_x = (int)(min_x + 0.5);
  mi_y = (int)(min_y + 0.5);

  ma_x = (int)(max_x - min_x + 0.5);
  ma_y = (int)(max_y - min_y + 0.5);
  static int exp_size[4];

  exp_size[0] = ma_x;
  exp_size[1] = ma_y;
  exp_size[2] = mi_x;
  exp_size[3] = mi_y;

  return exp_size;
}

// run function
int main(int argc, char** argv){
  int i,j;
  // power of expand
  int power = 10;

  // read original image
  Mat img, img_color;
  Mat img_R, img_G, img_B;
  vector<Mat> channels;

  img = imread("tokyoskytree.jpg", CV_LOAD_IMAGE_GRAYSCALE);
  img_color = imread("tokyoskytree.jpg");

  // split image
  split(img_color,channels);
  img_R = channels.at(0);
  img_G = channels.at(1);
  img_B = channels.at(2);


  // rotation angle
  double theta = -M_PI/6;
  // rotation matrix, inverse of rotation matrix
  float R_ro[2][2] = {{cos(theta), -sin(theta)}, {sin(theta), cos(theta)}};
  float R_ro_inv[2][2] = {{cos(theta), sin(theta)}, {-sin(theta), cos(theta)}};
  // expand matrix, inverse of expand matrix
  float R_expand[2][2] = {{power, 0}, {0, power}};
  float R_expand_inv[2][2] = {{1.0/power, 0}, {0, 1.0/power}};

  // calculate size of rotation image
  int *rotation_size = expand_size(R_ro, img.rows, img.cols);
  Mat img_rotation_R(rotation_size[0], rotation_size[1], CV_8U);
  Mat img_rotation_G(rotation_size[0], rotation_size[1], CV_8U);
  Mat img_rotation_B(rotation_size[0], rotation_size[1], CV_8U);

  // rotate image
  for(i=0; i<rotation_size[0]; i++)
    for(j=0; j<rotation_size[1]; j++){
      int co[2] = {i + rotation_size[2], j + rotation_size[3]};
      int *co_ = mat2x2(R_ro_inv, co);
      if ((co_[0] < 0) or (co_[1] < 0))
        continue;
      if ((co_[0] > img.rows) or (co_[1] > img.cols))
        continue;

      img_rotation_R.at<unsigned char>(i,j) = img_R.at<unsigned char>(co_[0],co_[1]);
      img_rotation_G.at<unsigned char>(i,j) = img_G.at<unsigned char>(co_[0],co_[1]);
      img_rotation_B.at<unsigned char>(i,j) = img_B.at<unsigned char>(co_[0],co_[1]);
    }

  // calculate size of expanded image
  int *exp_size = expand_size(R_expand, img_rotation_R.rows, img_rotation_R.cols);
  Mat img_expand_R(exp_size[0], exp_size[1], CV_8U);
  Mat img_expand_G(exp_size[0], exp_size[1], CV_8U);
  Mat img_expand_B(exp_size[0], exp_size[1], CV_8U);
  Mat img_expand_color(exp_size[0], exp_size[1],CV_8UC3);
  channels.at(0) = img_expand_R;
  channels.at(1) = img_expand_G;
  channels.at(2) = img_expand_B;

  // expand image
  for(i=0; i<img_expand_R.rows; i++)
    for(j=0; j<img_expand_R.cols; j++){
      int co[2] = {i, j};
      int *co_ = mat2x2(R_expand_inv, co);
      if ((co_[0] < 0) or (co_[1] < 0))
        continue;
      if ((co_[0] > img_rotation_R.rows) or (co_[1] > img_rotation_R.cols))
        continue;

      img_expand_R.at<unsigned char>(i,j) = img_rotation_R.at<unsigned char>(co_[0],co_[1]);
      img_expand_G.at<unsigned char>(i,j) = img_rotation_G.at<unsigned char>(co_[0],co_[1]);
      img_expand_B.at<unsigned char>(i,j) = img_rotation_B.at<unsigned char>(co_[0],co_[1]);
    }

  merge(channels, img_expand_color);

  // save image
  imwrite("result.jpg", img_expand_color);

  // show image
  namedWindow("Original", WINDOW_AUTOSIZE);
  imshow("Original", img_color);

  namedWindow("Expand", WINDOW_AUTOSIZE);
  imshow("Expand", img_expand_color);

  waitKey(0);
  return 0;
}
